### PR TITLE
Acd provide gps

### PR DESCRIPTION
**Pull request by Scumi**
ACD driver modified to provide GPS data to Air Control Display unit. The ACD makes use of the GPS data for the NEAREST function of the radio.

GPSState was changed to store VDOP and HDOP and the Parser to parse those values from the GPGSA sentence.

### CODE REVIEW

**Comment by Scumi** [comment]
> Those attributes need to be considered in `GPSState::Reset()`

done

**Comment by MaxKellermann** [comment]
The commit "NMEA/GPSState.hpp: Add VDOP and PDOP" still doesn't contain proper support code. This belongs in the same commit.

**Comment by Scumi** [comment]
Squashed the support code commits together so they can be cherry-picked.

File: src/NMEA/Info.cpp [hunk 1]
```diff
@@ -38,6 +38,9 @@ GPSState::Reset()
   satellites_used_available.Clear();
   satellite_ids_available.Clear();
   replay = false;
+  pdop = 0;
+  hdop = 0;
```
Comment (MaxKellermann) [inline src/NMEA/Info.cpp]: This line is unrelated to this commit.

File: src/NMEA/Info.cpp [hunk 2]
```diff
@@ -38,6 +38,9 @@ GPSState::Reset()
   satellites_used_available.Clear();
   satellite_ids_available.Clear();
   replay = false;
+  pdop = 0;
```
Comment (MaxKellermann) [inline src/NMEA/Info.cpp]: What does zero mean here? Is this some kind of "magic" value? Why isn't this documented?
(Don't be distracted from the fact that `hdop` isn't documented either - that's old and bad code. I just don't accept adding more equally bad code.)

File: src/NMEA/Info.cpp [hunk 3]
```diff
@@ -38,6 +38,9 @@ GPSState::Reset()
...
+  hdop = 0;
+  vdop = 0;
```
Comment (MaxKellermann) [inline src/NMEA/Info.cpp]: Same here: what does zero mean?

File: src/NMEA/Info.cpp (hunk 2 above)
Comment (hor63) [reply src/NMEA/Info.cpp]: Mmmh. It should be 1.0 as intial value. 0 is a bad initialization for that.
1.0 means that there is no dilution of precision, and is a much better default than 0 (no error at all, which is unrealistic)

HDOP, PDOP, and VDOP are well-defined terms in the NMEA specification. The only use of HDOP is in Logger/LoggerEPE where the meaning is epically described.
Here mentioning that the xDOP values are used for pass-through, and HDOP for error estimation should be sufficient.

Comment (Scumi) [reply src/NMEA/Info.cpp]: I agree, a value of 0 doesn't make sense and should not be used for initialization. 1.0 is not a good value either in my opinion, since that is the ideal state of this value (Ref.: https://en.wikipedia.org/wiki/Dilution_of_precision_(navigation)#Meaning_of_DOP_Values).

I would set it to 20 (or even better: 99). If you forward that value to other devices, they should discard it right away.

Comment (hor63) [reply src/NMEA/Info.cpp]: Better: -1 is an impossible value, and an indictor that no value is present. When sending it out leave the field empty.

Comment (MaxKellermann) [reply src/NMEA/Info.cpp]: @hor63 you mean like 2459e1e3378b7ca0a43e33ca0ce1ac759b97ae2f which I wrote 2 hours ago?

Comment (hor63) [reply src/NMEA/Info.cpp]: :D Exactly. @Scumi should build on that.

Comment (Scumi) [reply src/NMEA/Info.cpp]: Thanks for updating that code, Max. Updated mine to align with the new support code.

File: src/Logger/LoggerImpl.hpp [hunk 4]
```diff
@@ -71,6 +71,10 @@ class LoggerImpl
     bool satellites_used_available;
     /** GPS Horizontal Dilution of precision */
     double hdop;
+    /** GPS Position (3D) Dilution of precision */
+    double pdop;
```
Comment (MaxKellermann) [inline src/Logger/LoggerImpl.hpp]: Why copy this in the logger? The logger uses `hdop`, but doesn't use any of the new attributes (nor does any of your other commits add usage of it).

File: src/Device/Driver/AirControlDisplay.cpp [hunk 5]
```diff
@@ -123,6 +127,226 @@ ACDDevice::ParseNMEA(const char *_line, NMEAInfo &info)
     return false;
 }
 
+static void
```
Comment (MaxKellermann) [inline src/Device/Driver/AirControlDisplay.cpp]: Much of this is copy'n'pasted from somewhere else. Duplicate code is not acceptable.

File: src/Device/Driver/AirControlDisplay.cpp [hunk 6]
```diff
@@ -123,6 +127,226 @@ ACDDevice::ParseNMEA(const char *_line, NMEAInfo &info)
...
+           info.gps.satellite_ids[11],
+           info.gps.pdop,
+           info.gps.hdop,
+           info.gps.vdop);
+
+  return true;
+}
+
+void
+ACDDevice::OnCalculatedUpdate(const MoreData &basic,
```
Comment (MaxKellermann) [inline src/Device/Driver/AirControlDisplay.cpp]: Why implement `OnCalculatedUpdate()` when this doesn't actually uses the calculated data? Better implement `OnSensorUpdate()` instead.

File: src/Device/Driver/AirControlDisplay.cpp [hunk 7]
```diff
@@ -123,6 +127,226 @@ ACDDevice::ParseNMEA(const char *_line, NMEAInfo &info)
...
+                        const DerivedInfo &calculated)
+{
+  NullOperationEnvironment env;
+  char buffer[100];
+
+  if (FormatGPRMC(buffer, sizeof(buffer), basic))
```
Comment (MaxKellermann) [inline src/Device/Driver/AirControlDisplay.cpp]: Only send this if a new GPS location has been received; imagine one device has a 20 Hz pressure sensor, but doesn't provide location data; this would trigger sending those sentences 20 times a second.

File: NEWS.txt [hunk 8]
```diff
@@ -9,7 +9,7 @@ Version 7.0 - not yet released
   - parse wind from standard NMEA sentence WMV
   - driver for XC Tracer Vario
   - driver for KRT2 radio
-  - driver for Air Control Display altimeter
+  - driver for Air Control Display altimeter and radio
```
Comment (MaxKellermann) [inline NEWS.txt]: This should be part of the commit that adds the feature.

File: src/Device/Driver/AirControlDisplay.cpp (hunk 5 above)
Comment (Scumi) [reply src/Device/Driver/AirControlDisplay.cpp]: I agree, duplicate code makes everything a mess. Some of the code comes from another driver. As far as I understand this, I cannot use methods from another driver (which also rules out using NMEAOut to forward the data). Since each driver has to be maintained to what the device it is designed for can do, it might also be not practical to merge code and implement the methods. FLARM for example is an exception to that of course. 

Ideas for a way forward?

File: src/Device/Driver/AirControlDisplay.cpp (hunk 6 above)
Comment (Scumi) [reply src/Device/Driver/AirControlDisplay.cpp]: Hm, `OnSensorUpdate()` made the app freeze every few seconds. Might have to do with sending the NMEA sentences to the device without checking if a new GPS location has been received (see your other comment below). `OnCalculatedUpdate()` works fine, even if I don't need the calculated data.

File: src/Device/Driver/AirControlDisplay.cpp (hunk 7 above)
Comment (Scumi) [reply src/Device/Driver/AirControlDisplay.cpp]: I agree, makes sense.

Do you think I will still have a freezing app when would have a 20 Hz GPS receiver and use `OnSensorUpdate()`?

File: NEWS.txt (hunk 8 above)
Comment (Scumi) [reply NEWS.txt]: Radio driver would be complete with the addition of this PR. Will squash this commit into the "main" one.

**Comment by MaxKellermann** [comment]
There hasn't been any progress on fixing the problems for 2 months - let's close it until we have fixes.

**Comment by Scumi** [comment]
Yes, I had a bunch of questions that were not answered for 2 months. I can try to improve my code, although I cannot guarantee that I solve all issues you had with the code.
